recommendations for writing controllers

deads2k opened the pull request:
There have been questions about how to write controllers.  Many of the gotchas are not obvious to new authors.  This doc tries to help explain the common pitfalls.

@lavalamp @smarterclayton @kubernetes/rh-cluster-infra @kargakis ptal

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32778)

<!-- Reviewable:end -->

deads2k committed:
recommendations for writing controllers

sttts commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,101 @@
+# Writing Controllers
+
+A Kubernetes controller is simply an optimization on bash script that does repeated lists against the API server.
```
Why a bash script? :)

kargakis replied:
Is there any other doc that describes Kube controllers? If yes, link it here. If no, then this needs more info. Mention watching next to listing.

deads2k replied:
> Why a bash script? :)

Easy example that nearly everyone is familiar with and no one actually uses any other shell, right? :)

deads2k replied:
> Is there any other doc that describes Kube controllers? If yes, link it here. If no, then this needs more info. Mention watching next to listing.

Seems weird, but I don't know of any.

sttts commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,101 @@
+
+When you’re writing controllers, there are few guidelines that will help make sure you get the results and performance you’re looking expecting.
+
+1. Operate on one item at a time.  If you use a `workqueue.Interface`, you’ll be queue changes for a particular resource and later pop them in multiple “worker” gofuncs with a guarantee that no two gofuncs will work on the same item at the same time.
```
will be queue?

kargakis replied:
s/queue/queued/

sttts commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,101 @@
+ Distinct watches are updated independently.  Even with an objective ordering of “created resourceA/X” and “created resourceB/Y”, your controller could observe “created resourceB/Y” and “created resourceA/X”.
+
+
+1. Level driven, not edge driven.  Just like having a bash script that isn’t running all the time, your controller may off for an indeterminate amount of time before running again.
```
be off

sttts commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,101 @@
+
+1. Level driven, not edge driven.  Just like having a bash script that isn’t running all the time, your controller may off for an indeterminate amount of time before running again.
+
+ If an API object appears with a marker value of `true`, you can’t count on having seen it turn from `false` to `true`, only that you now observe it being `true`.  Even an API watch suffers from this problem, so be sure that you’re not counting on seeing a change unless your controller is also marking the information it last made the decision on in the object’s status.
```
"is also marking the information it last made the decision on in the object’s status." - I wished the English language had some more commas such that you don't need a non-regular grammar to parse it.

sttts commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,101 @@
+ You may see other mechanisms like reflectors and deltafifos driving controllers.  Those were older mechanisms that we later used to build the `SharedInformers`.  You should avoid using them in new controllers
+
+
+1. Never mutate original objects!  Caches are shared across controllers, this means that if you mutate your copy of an object, you’ll mess up other controllers (not just your own).  
```
it's actually not "your copy", that's what you are saying here ;-)

sttts commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,101 @@
+
+1. Percolate errors to the top level for consistent re-queuing.  We have a  `workqueue.RateLimitingInterface` to allow simple requeuing with reasonable backoffs.
+
+ Your main controller func should return an error when requeuing is necessary.  When it isn’t, it should se `utilruntime.HandleError` and return nil instead.  This makes it very easy for reviewers to inspect error handling cases and to be confident that your controller doesn’t accidentally lose things it should retry for.
```
se?

sttts commented on docs/devel/controllers.md:
````diff
@@ -0,0 +1,101 @@
+	return true
+}
+
+```
````
nit: some empty lines less

sttts commented:
Very helpful document!

kargakis commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,101 @@
+
+## Guidelines
+
+When you’re writing controllers, there are few guidelines that will help make sure you get the results and performance you’re looking expecting.
```
s/looking expecting/looking for/ or s/looking expecting/expecting/

kargakis commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,101 @@
+
+1. Operate on one item at a time.  If you use a `workqueue.Interface`, you’ll be queue changes for a particular resource and later pop them in multiple “worker” gofuncs with a guarantee that no two gofuncs will work on the same item at the same time.
+
+ Many controllers must trigger off multiple resources (I need to check X if Y changes), but nearly all controllers can collapse those into a queue of “check this X” based on relationships.  For instance, a ReplicaSetController needs to react to a pod being deleted, but it does that by finding the related ReplicaSets and queuing those. 
```
Add quotes in the parenthesis

kargakis commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,101 @@
+
+1. Watches and Informers will “sync”.  Periodically, they will deliver every matching object in the cluster to your `Update` method.  This is good for cases where you may need to take additional action on the object, but sometimes you know there won’t be more work to do.
+
+ In cases where you are *certain* that you always requeue failing items, you can compare the resource version of the old and new objects.  If they are the same, you skip requeuing the work.
```
"failing items" doesn't make any sense to me.  How about "In case where you are certain that you don't need to requeue items with no new changes" or something similar

kargakis commented:
NIce idea

kargakis commented:
cc: @bparees

deads2k commented:
Fixed line breaks and phrasing comments.  I squashed this one since they were minor and the whitespace made the diff noisy.  I'll keep further comments split out for easier review.

bparees commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,164 @@
+ `Update` method.  This is good for cases where you may need to take additional action on the object, but sometimes you
+ know there won’t be more work to do.
+
+ In cases where you are *certain* that you don't need to requeue items when no new changes, you can compare the
```
"when no new changes" -> when there are no new changes?

DirectXMan12 commented:
In the example, it might not be a bad idea to show how you get items into the queue initially.

ravilr commented:
this is a very good doc on how to write a k8s controller. Can we get this merged? Thanks.

deads2k commented:
> this is a very good doc on how to write a k8s controller. Can we get this merged? Thanks.

@lavalamp Can you take a pass on this?

lavalamp commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,164 @@
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
```
Add comment

lavalamp replied:
Actually I would comment everything in the example snippet heavily.

lavalamp commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,164 @@
+
+# Writing Controllers
+
+A Kubernetes controller is simply an optimization on shell script that does repeated lists against the API server.
```
A Kubernetes controller is an active reconciliation process. That is, it watches some object for the world's desired state, and it watches the world's actual state, too. Then, it sends instructions to try and make the world's current state be more like the desired state.

The simplest implementation of this is a loop:

```go
for {
  desired := getDesiredState()
  current := getCurrentState()
  makeChanges(desired, current)
}
```

Watches, etc, are all merely optimizations of this logic.

lavalamp commented on docs/devel/controllers.md:
```diff
@@ -0,0 +1,164 @@
+ only that you now observe it being `true`.  Even an API watch suffers from this problem, so be sure that you’re not
+ counting on seeing a change unless your controller is also marking the information it last made the decision on in
+ the object's status.
+
```
Don't forget that the current state may change at any moment--it's not sufficient to just watch the desired state. If you use the absence of objects in the desired state to indicate that things in the current state should be deleted, make sure you don't have a bug in your observation code (e.g., act before your cache has filled).

kargakis replied:
Made me think that we should have a spot for Generation/ObservedGeneration. Controller authors should have ObservedGeneration updated in order to notify clients that their changes were observed.

lavalamp reviewed (commented):
I have some suggested additions/rewrites.

lavalamp commented:
Oh, good point. Controllers should save observed version somewhere!

On Wed, Oct 12, 2016 at 3:05 PM, Michail Kargakis <notifications@github.com>
wrote:

> *@kargakis* commented on this pull request.
> ------------------------------
>
> In docs/devel/controllers.md
> <https://github.com/kubernetes/kubernetes/pull/32778>:
>
> > +
> +1. Random ordering between resources. When controllers queue off multiple types of resources, there is no guarantee
> + of ordering amongst those resources.
> +
> + Distinct watches are updated independently.  Even with an objective ordering of “created resourceA/X” and “created
> + resourceB/Y”, your controller could observe “created resourceB/Y” and “created resourceA/X”.
> +
> +
> +1. Level driven, not edge driven.  Just like having a shell script that isn’t running all the time, your controller
> + may be off for an indeterminate amount of time before running again.
> +
> + If an API object appears with a marker value of `true`, you can’t count on having seen it turn from `false` to `true`,
> + only that you now observe it being `true`.  Even an API watch suffers from this problem, so be sure that you’re not
> + counting on seeing a change unless your controller is also marking the information it last made the decision on in
> + the object's status.
> +
>
> Made me think that we should have a spot for
> Generation/ObservedGeneration. Controller authors should have
> ObservedGeneration updated in order to notify clients that their changes
> were observed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32778>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglldJTh65iXSUcojYuYWgppuXR0Hwks5qzVmvgaJpZM4J9xSc>
> .
>

deads2k commented:
Most comments addressed.



> Made me think that we should have a spot for Generation/ObservedGeneration. Controller authors should have ObservedGeneration updated in order to notify clients that their changes were observed.

Can we add this later?  There are some cases where generation/observedgeneration are extremely useful, but there are also cases where they aren't.  I don't have a crisp rule in my head for deciding one case from another.  It's mostly a case of, "I'll know it when I see it", which is largely unhelpful in a document like this.

I think everything else here can be useful and separate pull could add a crisp rule.  Alternatively, if either of you have an easy rule in your heads, I'd be happy to add it.

kargakis commented:
Not sure if you mean that they are not useful for some cases per se or that
it wouldn't be useful to have ObservedGeneration updated in those cases. I
agree on the latter.

On Thu, Oct 13, 2016 at 3:20 PM, David Eads <notifications@github.com>
wrote:

> Most comments addressed.
>
> Made me think that we should have a spot for
> Generation/ObservedGeneration. Controller authors should have
> ObservedGeneration updated in order to notify clients that their changes
> were observed.
>
> Can we add this later? There are some cases where
> generation/observedgeneration are extremely useful, but there are also
> cases where they aren't. I don't have a crisp rule in my head for deciding
> one case from another. It's mostly a case of, "I'll know it when I see it",
> which is largely unhelpful in a document like this.
>
> I think everything else here can be useful and separate pull could add a
> crisp rule. Alternatively, if either of you have an easy rule in your
> heads, I'd be happy to add it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32778#issuecomment-253510728>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFf_5VN5NtH5X9OUyIJXGcObFsVV4xks5qzjAWgaJpZM4J9xSc>
> .
>

lavalamp commented:
I don't feel that strongly, but I don't think there's any case where it's
*harmful*...

On Thu, Oct 13, 2016 at 6:20 AM, David Eads <notifications@github.com>
wrote:

> Most comments addressed.
>
> Made me think that we should have a spot for
> Generation/ObservedGeneration. Controller authors should have
> ObservedGeneration updated in order to notify clients that their changes
> were observed.
>
> Can we add this later? There are some cases where
> generation/observedgeneration are extremely useful, but there are also
> cases where they aren't. I don't have a crisp rule in my head for deciding
> one case from another. It's mostly a case of, "I'll know it when I see it",
> which is largely unhelpful in a document like this.
>
> I think everything else here can be useful and separate pull could add a
> crisp rule. Alternatively, if either of you have an easy rule in your
> heads, I'd be happy to add it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32778#issuecomment-253510728>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglgoWcsbaKXDA5SMtQLiqWvzBdAHiks5qzjAFgaJpZM4J9xSc>
> .
>

kargakis commented:
Adding lgtm, @deads2k will open a separate issue about documenting observedGeneration

deads2k commented:
@kargakis opened https://github.com/kubernetes/kubernetes/issues/35036 .
